drivers/interrupt_controller/intc_plic.c

```diff
@@ -20,24 +20,76 @@
 #include <zephyr/sw_isr_table.h>
 #include <zephyr/drivers/interrupt_controller/riscv_plic.h>
 #include <zephyr/irq.h>
+#include <zephyr/sys/__assert.h>
 
-#define PLIC_MAX_PRIO	DT_INST_PROP(0, riscv_max_priority)
-#define PLIC_PRIO	DT_INST_REG_ADDR_BY_NAME_U64(0, prio)
-#define PLIC_IRQ_EN	DT_INST_REG_ADDR_BY_NAME_U64(0, irq_en)
-#define PLIC_REG	DT_INST_REG_ADDR_BY_NAME_U64(0, reg)
+#define PLIC_BASE_ADDR(n) DT_INST_REG_ADDR(n)
+#define PLIC_REG_PRIO_OFFSET 0x0
+#define PLIC_REG_TRIG_TYPE_OFFSET 0x1080
+#define PLIC_REG_IRQ_EN_OFFSET 0x2000
+#define PLIC_REG_REGS_OFFSET 0x200000
 
-#define PLIC_IRQS        (CONFIG_NUM_IRQS - CONFIG_2ND_LVL_ISR_TBL_OFFSET)
-#define PLIC_EN_SIZE     ((PLIC_IRQS >> 5) + 1)
-
-#define PLIC_EDGE_TRIG_TYPE   (PLIC_MAX_PRIO + DT_INST_PROP(0, riscv_trigger_reg_offset))
 #define PLIC_EDGE_TRIG_SHIFT  5
 
 struct plic_regs_t {
 	uint32_t threshold_prio;
 	uint32_t claim_complete;
 };
 
+typedef void (*riscv_plic_irq_config_func_t)(void);
+struct plic_config {
+	mem_addr_t prio;
+	mem_addr_t irq_en;
+	mem_addr_t reg;
+	mem_addr_t trig;
+	uint32_t max_prio;
+	uint32_t num_irqs;
+	uint32_t isr_offset;
+	riscv_plic_irq_config_func_t irq_config_func;
+};
+
+#ifdef CONFIG_XIP
+static const struct plic_config plic_config_0;
+#define PLIC_INIT_FN_ARG(arg)
+#else
+#define PLIC_INIT_FN_ARG(arg) const struct device *arg
+#endif
+
 static int save_irq;
+static struct device *save_dev;
```

Comment:
Don't do this - if there is instance-specific data, encapsulate it in a `struct plic_data`.

Reply:
`save_irq` & `save_dev` record the IRQ line & instance that is active for the current CPU, they should be per CPU instead of per intc instance